// takeshape-ssg/lib/config.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.prepareConfig = prepareConfig;
exports.loadConfig = loadConfig;

var _jsYaml = _interopRequireDefault(require("js-yaml"));

var _fs = _interopRequireDefault(require("fs"));

var _path = _interopRequireDefault(require("path"));

var _jsonVariables = _interopRequireDefault(require("json-variables"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; var ownKeys = Object.keys(source); if (typeof Object.getOwnPropertySymbols === 'function') { ownKeys = ownKeys.concat(Object.getOwnPropertySymbols(source).filter(function (sym) { return Object.getOwnPropertyDescriptor(source, sym).enumerable; })); } ownKeys.forEach(function (key) { _defineProperty(target, key, source[key]); }); } return target; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

function prepareConfig(configFilePath, config, envConfig = {}) {
  const cwd = _path.default.dirname(configFilePath);

  const combined = _objectSpread({
    env: {},
    context: {},
    routes: {},
    imageBaseUrl: 'https://images.takeshape.io',
    assetBaseUrl: 'https://assets.takeshape.io',
    imageDefaults: {
      auto: 'compress,format'
    },
    locale: 'en-us',
    pathPrefix: ''
  }, config, {
    configFilePath,
    cwd,
    templatePath: _path.default.join(cwd, config.templatePath || 'templates'),
    buildPath: _path.default.join(cwd, config.buildPath || 'build'),
    staticPath: _path.default.join(cwd, config.staticPath || 'static'),
    dates: _objectSpread({
      tz: 'America/New_York',
      format: 'LLL'
    }, config.dates),
    usageStats: false
  }, envConfig);

  return (0, _jsonVariables.default)(combined, {
    heads: '${',
    tails: '}'
  });
}

function loadConfig(configFilePath, envConfig) {
  try {
    const config = _jsYaml.default.load(_fs.default.readFileSync(configFilePath, 'utf8'));

    return prepareConfig(configFilePath, config, envConfig);
  } catch (e) {
    throw new Error(`Failed to load tsg config from ${configFilePath}`);
  }
}
