// takeshape-routing/lib/urls.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.getPath = getPath;
exports.getAssetUrl = getAssetUrl;
exports.getImageUrl = getImageUrl;
exports.assetBaseUrl = exports.imageBaseUrl = void 0;

var _queryString = _interopRequireDefault(require("query-string"));

var _pickBy = _interopRequireDefault(require("lodash/pickBy"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

const imageBaseUrl = 'https://images.takeshape.io';
exports.imageBaseUrl = imageBaseUrl;
const assetBaseUrl = 'https://assets.takeshape.io';
exports.assetBaseUrl = assetBaseUrl;

function getPath(key) {
  return typeof key === 'object' ? 'path' in key ? key.path : key.s3Key : key;
}

function formatQuery(query) {
  if (query) {
    const filtered = (0, _pickBy.default)(query, value => value !== '');

    if (Object.keys(filtered).length) {
      return '?' + _queryString.default.stringify(filtered);
    }
  }

  return '';
}

function escapePath(path) {
  return path.split('/').map(encodeURIComponent).join('/');
}

function getAssetUrl(s3Key, options = {}) {
  const path = getPath(s3Key);
  const baseUrl = options.baseUrl || assetBaseUrl;
  return path ? `${baseUrl}/${escapePath(path)}` : '';
}

function getImageUrl(key, query, options = {}) {
  const path = getPath(key);
  const baseUrl = options.baseUrl || imageBaseUrl;
  return path ? `${baseUrl}/${escapePath(path)}${formatQuery(query)}` : '';
}
