// takeshape-cli/lib/commands/init/messages.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.warning = warning;
exports.info = info;
exports.forgotPasswordWarning = forgotPasswordWarning;
exports.noProjectsError = noProjectsError;
exports.glitchConfirm = exports.consoleWarning = exports.noSitesConfiguredWarning = exports.noSiteWarning = exports.writingFilesMsg = exports.missingPasswordMsg = exports.invalidLoginMsg = void 0;

var _chalk = _interopRequireDefault(require("chalk"));

var _log = _interopRequireDefault(require("../../log"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function warning(msg) {
  return () => msg.split('\n').forEach(line => (0, _log.default)(_chalk.default.black.bgYellow(line)));
}

function info(msg) {
  return () => (0, _log.default)(msg);
}

const invalidLoginMsg = info('Invalid email/password combination');
exports.invalidLoginMsg = invalidLoginMsg;
const missingPasswordMsg = info('No password set in TakeShape. Follow the forgot password flow to set a password.');
exports.missingPasswordMsg = missingPasswordMsg;
const writingFilesMsg = info('Writing .tsgrc, .graphqlconfig.');
exports.writingFilesMsg = writingFilesMsg;
const noSiteWarningMsg = `You will be able to develop locally but will be \
unable to deploy until you configure a static site. Don't worry though you'll be able to \
configure a static site later on and we'll remind you again when you try to deploy.`;
const noSiteWarning = info(noSiteWarningMsg);
exports.noSiteWarning = noSiteWarning;
const noSitesConfiguredWarning = info(`No sites configured. ${noSiteWarningMsg}`);
exports.noSitesConfiguredWarning = noSitesConfiguredWarning;
const consoleWarning = warning('**************************************************************************************************\n' + ' DO NOT COMMIT THESE FILES! ADD THESE FILES TO YOUR VCS IGNORE FILE.                              \n' + '                                                                                                  \n' + '   .tsgrc                                                                                         \n' + '   .graphqlconfig                                                                                 \n' + '                                                                                                  \n' + ' DO NOT COMMIT THESE FILES! ADD THESE FILES TO YOUR VCS IGNORE FILE.                              \n' + '**************************************************************************************************');
exports.consoleWarning = consoleWarning;

function forgotPasswordWarning(params) {
  (0, _log.default)(`If you forgot your credentials please visit ${params.appUrl}/forgot-password`);
}

function noProjectsError(params) {
  (0, _log.default)(`No projects configured set one up: ${params.appUrl}/projects`);
}

const glitchConfirm = "We detected you're running in Glitch. Is this accurate?";
exports.glitchConfirm = glitchConfirm;
